lib/bundling.js

```diff
@@ -0,0 +1,326 @@
+      var index = 0;
+      this._data.forEach(function(data) {
+        var endIndex = index + data.elements.length;
+        if (endIndex > responseValue.length) {
+          console.warn('subresponse size is not enough.');
+          return;
+        }
+        subresponses.push(responseValue.slice(index, endIndex));
+        index = endIndex;
+      });
+    }
+    for (var i = 0; i < this._data.length; ++i) {
+      if (this._data[i].emitter) {
+        this._data[i].emitter._callback(err, subresponses[i]);
+      }
+    }
+  }.bind(this));
+};
+
+/**
+ * Appends the list of elements into the task.
+ * @param {Object[]} elements - the new list of elements.
+ * @param {BundleEventEmitter} emitter - the event emitter.
+ */
+Task.prototype.extend = function(elements, emitter) {
+  var bytes = 0;
+  elements.forEach(function(element) {
+    bytes += this._byteLengthFunction(element);
+  }.bind(this));
+  this._data.push({
+    'elements': elements,
+    'bytes': bytes,
+    'emitter': emitter
+  });
+};
+
+/**
+ * Cancels a part of elements.
+ * @param {String} id - The identifier of the part of elements.
+ */
+Task.prototype.cancel = function(eventId) {
+  for (var i = 0; i < this._data.length; ++i) {
+    if (this._data[i].emitter && this._data[i].emitter._eventId === eventId) {
+      this._data.splice(i, 1);
+      return true;
+    }
+  }
+  return false;
+};
+
+/**
+ * Organizes requests for an api service that requires to bundle them.
+ *
+ * @param {BundleOptions} bundleOptions - configures strategy this instance
+ *   uses when executing bundled functions.
+ * @param {BundleDescriptor} bundleDescriptor - the description of the bundling.
+ * @constructor
+ */
+function BundleExecutor(bundleOptions, bundleDescriptor) {
+  this._options = bundleOptions;
+  this._descriptor = bundleDescriptor;
+  this._tasks = {};
+  this._timers = {};
+  this._events = {};
+}
+
+exports.BundleExecutor = BundleExecutor;
+
+/**
+ * Schedule a method call.
+ *
+ * @param {function} apiCall - the function for an API call.
+ * @param {Object} request - the request object to be bundled with others.
+ * @param {APICallback} callback - the callback to be called when the API call
+ *   has finished.
+ */
+BundleExecutor.prototype.schedule = function(
+    apiCall, request, emitter) {
+  var bundleId = computeBundleId(
+      request, this._descriptor.requestDiscriminatorFields);
+  if (bundleId === undefined) {
+    var err = new Error('request object ' + JSON.stringify(request) +
+                        ' is not compatible with discriminator fields');
+    emitter._callback(err);
+  }
+
+  if (!(bundleId in this._tasks)) {
+    this._tasks[bundleId] = new Task(
+        apiCall, request, this._descriptor.bundledField,
+        this._descriptor.subresponseField, this._descriptor.byteLengthFunction);
+  }
+  var task = this._tasks[bundleId];
+  this._events[emitter._eventId] = bundleId;
+
+  task.extend(request[this._descriptor.bundledField], emitter);
+
```

Comment:
optional: handle limits, in addition to thresholds. This can also be in a follow-up PR, but let's create a tracking issue if so.